weed/server/volume_server_handlers.go

```diff
@@ -56,20 +57,29 @@ func (vs *VolumeServer) privateStoreHandler(w http.ResponseWriter, r *http.Reque
 		vs.guard.WhiteList(vs.DeleteHandler)(w, r)
 	case "PUT", "POST":
 
-		// wait until in flight data is less than the limit
 		contentLength := getContentLength(r)
-
 		// exclude the replication from the concurrentUploadLimitMB
-		if vs.concurrentUploadLimit != 0 && r.URL.Query().Get("type") != "replicate" &&
-			atomic.LoadInt64(&vs.inFlightUploadDataSize) > vs.concurrentUploadLimit {
-			err := fmt.Errorf("reject because inflight upload data %d > %d", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
-			glog.V(1).Infof("too many requests: %v", err)
-			writeJsonError(w, r, http.StatusTooManyRequests, err)
-			return
+		if r.URL.Query().Get("type") != "replicate" && vs.concurrentUploadLimit != 0 {
+			startTime := time.Now()
+			vs.inFlightUploadDataLimitCond.L.Lock()
+			for vs.inFlightUploadDataSize > vs.concurrentUploadLimit {
+				//wait timeout check
+				if startTime.Add(vs.inflightUploadDataTimeout).Before(time.Now()) {
+					vs.inFlightUploadDataLimitCond.L.Unlock()
+					err := fmt.Errorf("reject because inflight upload data %d > %d, and wait timeout", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
+					glog.V(1).Infof("too many requests: %v", err)
+					writeJsonError(w, r, http.StatusTooManyRequests, err)
+					return
+				}
+				glog.V(4).Infof("wait because inflight upload data %d > %d", vs.inFlightUploadDataSize, vs.concurrentUploadLimit)
+				vs.inFlightUploadDataLimitCond.Wait()
+			}
+			vs.inFlightUploadDataLimitCond.L.Unlock()
 		}
 		atomic.AddInt64(&vs.inFlightUploadDataSize, contentLength)
 		defer func() {
 			atomic.AddInt64(&vs.inFlightUploadDataSize, -contentLength)
+			vs.inFlightUploadDataLimitCond.Signal()
```

Comment:
This also need to be conditional?

Reply:
It does not need.
<img width="518" alt="image" src="https://user-images.githubusercontent.com/28077875/173987072-d0d82fdb-2e62-4778-bd3f-2c07f506dc43.png">

Reply:
It's still an unnecessary call to an c library.

Reply:
Got it.
`if r.URL.Query().Get("type") != "replicate" && vs.concurrentUploadLimit != 0 ` or `if vs.concurrentUploadLimit != 0`， which condition  is better?

Reply:
modified, added `if vs.concurrentUploadLimit != 0 {`